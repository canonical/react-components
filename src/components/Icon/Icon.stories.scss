@import "~vanilla-framework/scss/settings";
@import "~vanilla-framework/scss/base_placeholders";
@import "~vanilla-framework/scss/base_icon-definitions";
@import "~vanilla-framework/scss/patterns_icons";

@include vf-b-placeholders;
@include vf-p-icons-common;
@include vf-p-icon-containers;

.p-icon--custom {
  @extend %icon;

  background-image: url(#{$assets-path}a4d31b28-icon-quote.svg);
}
