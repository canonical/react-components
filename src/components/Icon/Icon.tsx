import classNames from "classnames";
import type { HTMLProps } from "react";
import React from "react";

import type { ClassName, PropsWithSpread, ValueOf } from "types";

export const ICONS = {
  anchor: "anchor",
  chevronDown: "chevron-down",
  chevronUp: "chevron-up",
  close: "close",
  code: "code",
  collapse: "collapse",
  copy: "copy",
  delete: "delete",
  drag: "drag",
  error: "error",
  expand: "expand",
  externalLink: "external-link",
  help: "help",
  information: "information",
  menu: "menu",
  minus: "minus",
  plus: "plus",
  search: "search",
  share: "share",
  spinner: "spinner",
  success: "success",
  user: "user",
  warning: "warning",
  facebook: "facebook",
  twitter: "twitter",
  instagram: "instagram",
  linkedin: "linkedin",
  youtube: "youtube",
  github: "github",
  rss: "rss",
  email: "email",
} as const;

export type Props = PropsWithSpread<
  {
    /**
     * Optional classes to add to the icon element.
     */
    className?: ClassName;
    /**
     * Whether to show the light variant of the icon.
     */
    light?: boolean;
    /**
     * The name of the icon.
     */
    name: ValueOf<typeof ICONS> | string;
  },
  HTMLProps<HTMLElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Icon](https://docs.vanillaframework.io/patterns/icons/).
 *
 * Icons provide visual context and enhance usability.
 *
 * @param name One of built-in Vanilla icons or a name of a custom icon that follows `p-icon--{name}` convention.
 * @returns Icon
 */
const Icon = ({ className, light, name, ...props }: Props): JSX.Element => (
  <i
    className={classNames(className, `p-icon--${name}`, {
      "is-light": light,
    })}
    {...props}
  />
);

export default Icon;
