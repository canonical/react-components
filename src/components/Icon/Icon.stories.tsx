import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Icon, { ICONS } from "./Icon";

import "./Icon.stories.scss";

const meta: Meta<typeof Icon> = {
  component: Icon,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Icon>;

export const Base: Story = {
  name: "Base",

  args: {
    name: "facebook",
  },
};

export const Default: Story = {
  render: () => <Icon name={ICONS.plus} />,
  name: "Default",
};

/**
 * To use custom icons that are not included in Vanilla you need to provide your own icon styling following the `.p-icon--{name}` convention.
 */
export const Custom: Story = {
  render: () => <Icon name="custom" />,
  name: "Custom",
};

/**
 * To use custom icons that provide the name of the social media icon following the `.p-icon--{name}` convention.
 */
export const Social: Story = {
  render: () => <Icon name={ICONS.facebook} />,
  name: "Social",
};
