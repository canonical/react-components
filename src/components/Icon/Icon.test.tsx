import { render } from "@testing-library/react";
import React from "react";

import Icon, { ICONS } from "./Icon";

describe("Icon", () => {
  it("renders", () => {
    const { container } = render(<Icon name={ICONS.success} />);
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toMatchSnapshot();
  });

  it("sets correct class name based on given name", () => {
    const { container } = render(<Icon name="test" />);
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toHaveClass("p-icon--test");
  });

  it("can be given a custom class name", () => {
    const { container } = render(<Icon className="custom-class" name="test" />);
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toHaveClass("custom-class p-icon--test");
  });

  it("can be given standard HTML props", () => {
    const style = { width: "200px" };
    const { container } = render(<Icon name="test" style={style} />);
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toHaveAttribute("style", "width: 200px;");
  });
});
