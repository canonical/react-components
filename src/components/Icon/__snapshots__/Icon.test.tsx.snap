// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders 1`] = `
<i
  class="p-icon--success"
/>
`;
